client/client.go

```diff
@@ -365,23 +445,47 @@ func (c *client) handleDispatcher(loopCtx context.Context, dc string, tsoDispatc
 		// we will try to create the stream first.
 		if stream == nil {
 			if needUpdate {
-				err = c.updateLeader()
+				err = c.updateMember()
 				if err != nil {
 					select {
 					case <-loopCtx.Done():
 						return
 					default:
 					}
-					log.Error("[pd] failed updateLeader", errs.ZapError(err))
+					log.Error("[pd] failed update member", errs.ZapError(err))
 					continue
 				}
 				needUpdate = false
 			}
-			ctx, cancel = context.WithCancel(loopCtx)
-			done := make(chan struct{})
-			go c.checkStreamTimeout(ctx, cancel, done)
-			stream, err = pdpb.NewPDClient(c.getClientConnByDCLocation(dc)).Tso(ctx)
-			done <- struct{}{}
+			cc, url := c.getClientConnByDCLocation(dc)
+			cctx, cancel = context.WithCancel(loopCtx)
+			stream, err = c.createTsoStream(cctx, cancel, pdpb.NewPDClient(cc))
+			failpoint.Inject("unreachableNetwork", func() {
+				cancel()
+				stream, cancel = nil, nil
+				err = status.New(codes.Unavailable, "unavailable").Err()
+			})
+			rpcErr, ok := status.FromError(err)
+			// encounter the network error
+			if ok && isNetworkError(rpcErr.Code()) {
+				log.Info("fall back to use follower to redirect tso stream")
+				followerClient := c.followerClient()
+				if followerClient != nil {
+					// create the follower stream
+					cctx, cancel = context.WithCancel(loopCtx)
```

Comment:
I think we can't directly cover the `cancel` here which will cause the goroutine leak by the previous cancel in line461.

Reply:
I think we don't need to create a new context by`cctx, cancel = context.WithCancel(loopCtx)` here. We can directly pass the previous `cctx` which means `checkAllocator` don't need to handnle the redirectCancel. Then the `cancel` only be reference with 1 goroutine and it will be solved in the following error handling.